> Dependency Tree is empty

Hi,

thanks for the plugin, which I tried to use.
My app does not (yet) have a separate app folder structure and therefore only one build.gradle.

So I added plugins (https://github.com/nextcloud/android/pull/5733) and ran
./gradlew generateModulesGraphvizText -Pmodules.graph.print.statistics=true

which then gave me:
```
> Task :generateModulesGraphvizText FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':generateModulesGraphvizText'.
> Dependency Tree is empty
```

Thank you!

--- a/example/core/build.gradle
+++ b/example/core/build.gradle
@@ -1,4 +1,12 @@
+plugins {
+  id('com.jraska.module.graph.assertion')
+}
+
 apply plugin: 'java-library'
 
 dependencies {
+}
+
+moduleGraphAssert {
+  maxHeight = 0
 }

--- a/plugin/src/main/kotlin/com/jraska/module/graph/DependencyGraph.kt
+++ b/plugin/src/main/kotlin/com/jraska/module/graph/DependencyGraph.kt
@@ -1,6 +1,6 @@
 package com.jraska.module.graph
 
-class DependencyGraph() {
+class DependencyGraph private constructor() {
   private val nodes = mutableMapOf<String, Node>()
 
   fun findRoot(): Node {
@@ -56,7 +56,9 @@ class DependencyGraph() {
   }
 
   fun subTree(key: String): DependencyGraph {
-    val dependencyTree = DependencyGraph()
+    require(nodes.contains(key)) { "Dependency Tree doesn't contain module: $key" }
+
+    val dependencyTree = createSingular(key)
 
     addConnections(nodes.getValue(key), dependencyTree)
 
@@ -110,7 +112,19 @@ class DependencyGraph() {
   }
 
   companion object {
+    fun createSingular(singleModule: String): DependencyGraph {
+      val dependencyGraph = DependencyGraph()
+
+      dependencyGraph.getOrCreate(singleModule)
+
+      return dependencyGraph
+    }
+
     fun create(dependencies: List<Pair<String, String>>): DependencyGraph {
+      if (dependencies.isEmpty()) {
+        throw IllegalArgumentException("Graph cannot be empty. Use createSingular for cases with no dependencies")
+      }
+
       val graph = DependencyGraph()
       dependencies.forEach { graph.addEdge(it.first, it.second) }
       return graph

--- a/plugin/src/main/kotlin/com/jraska/module/graph/GraphvizWriter.kt
+++ b/plugin/src/main/kotlin/com/jraska/module/graph/GraphvizWriter.kt
@@ -11,7 +11,13 @@ object GraphvizWriter {
 
     stringBuilder.append("digraph G {\n")
 
-    dependencyGraph.dependencyPairs()
+    val dependencyPairs = dependencyGraph.dependencyPairs()
+    if(dependencyPairs.isEmpty()) {
+      stringBuilder.append("\"${dependencyGraph.findRoot().key}\"")
+        .append("\n")
+    }
+
+    dependencyPairs
       .forEach { connection ->
         stringBuilder.append("\"${connection.first}\"")
           .append(" -> ")

--- a/plugin/src/main/kotlin/com/jraska/module/graph/assertion/GradleDependencyGraphFactory.kt
+++ b/plugin/src/main/kotlin/com/jraska/module/graph/assertion/GradleDependencyGraphFactory.kt
@@ -7,7 +7,11 @@ import org.gradle.api.internal.artifacts.dependencies.DefaultProjectDependency
 object GradleDependencyGraphFactory {
 
   fun create(project: Project, configurationsToLook: Set<String>): DependencyGraph {
-    val dependencies = project.listDependencyPairs(configurationsToLook)
+    val dependencies = project.listAllDependencyPairs(configurationsToLook)
+
+    if (dependencies.isEmpty()) {
+      return DependencyGraph.createSingular(project.moduleDisplayName())
+    }
 
     val fullDependencyGraph = DependencyGraph.create(dependencies)
 
@@ -18,7 +22,7 @@ object GradleDependencyGraphFactory {
     }
   }
 
-  private fun Project.listDependencyPairs(configurationsToLook: Set<String>): List<Pair<String, String>> {
+  private fun Project.listAllDependencyPairs(configurationsToLook: Set<String>): List<Pair<String, String>> {
     return rootProject.subprojects
       .flatMap { project ->
         project.configurations

--- a/plugin/src/test/kotlin/com/jraska/module/graph/DependencyGraphTest.kt
+++ b/plugin/src/test/kotlin/com/jraska/module/graph/DependencyGraphTest.kt
@@ -62,6 +62,39 @@ class DependencyGraphTest {
   }
 
   @Test
+  fun subtreeOfLeafModuleIsNotEmpty() {
+    val dependencyTree = DependencyGraph.create(
+      "feature" to "lib",
+      "lib" to "core",
+      "app" to "feature",
+      "feature" to "core",
+      "app" to "core"
+    )
+
+    assert(dependencyTree.subTree("core").findRoot().key == "core")
+  }
+
+  @Test
+  fun singleModuleTreeWorks() {
+    val dependencyGraph = DependencyGraph.createSingular(":app")
+
+    assert(dependencyGraph.dependencyPairs().isEmpty())
+    assert(dependencyGraph.height() == 0)
+    assert(dependencyGraph.statistics().modulesCount == 1)
+    assert(dependencyGraph.statistics().edgesCount == 0)
+    assert(dependencyGraph.statistics().height == 0)
+    assert(dependencyGraph.statistics().height == 0)
+    assert(dependencyGraph.findRoot().key == ":app")
+    assert(dependencyGraph.heightOf(":app") == 0)
+    assert(dependencyGraph.longestPath().nodeNames.equals(listOf(":app")))
+  }
+
+  @Test(expected = IllegalArgumentException::class)
+  fun cannotCreateEmptyGraph() {
+    DependencyGraph.create()
+  }
+
+  @Test
   fun countsStatisticsWell() {
     val dependencyTree = DependencyGraph.create(
       ":app" to ":core",

--- a/plugin/src/test/kotlin/com/jraska/module/graph/assertion/GradleDependencyGraphFactorySingleModuleProjectTest.kt
+++ b/plugin/src/test/kotlin/com/jraska/module/graph/assertion/GradleDependencyGraphFactorySingleModuleProjectTest.kt
@@ -1,0 +1,54 @@
+package com.jraska.module.graph.assertion
+
+import com.jraska.module.graph.GraphvizWriter
+import org.gradle.api.internal.project.DefaultProject
+import org.gradle.api.plugins.JavaLibraryPlugin
+import org.gradle.testfixtures.ProjectBuilder
+import org.junit.Before
+import org.junit.Test
+
+class GradleDependencyGraphFactorySingleModuleProjectTest {
+
+  private val EXPECTED_SINGLE_MODULE = """digraph G {
+":app"
+}"""
+
+  private val EXPECTED_ROOT_MODULE = """digraph G {
+"root  some-root"
+}"""
+
+  private lateinit var singleModule: DefaultProject
+  private var rootProject: DefaultProject? = null
+
+  @Before
+  fun setUp() {
+    rootProject = createProject("some-root")
+    singleModule = createProject("app")
+  }
+
+  @Test
+  fun generatesProperGraph() {
+    val dependencyGraph = GradleDependencyGraphFactory.create(singleModule, Api.API_IMPLEMENTATON_CONFIGURATIONS)
+
+    val graphvizText = GraphvizWriter.toGraphviz(dependencyGraph)
+    assert(EXPECTED_SINGLE_MODULE == graphvizText)
+  }
+
+  @Test
+  fun generatesProperGraphOnRootModule() {
+    val dependencyGraph = GradleDependencyGraphFactory.create(rootProject!!, Api.API_IMPLEMENTATON_CONFIGURATIONS)
+
+    val graphvizText = GraphvizWriter.toGraphviz(dependencyGraph)
+    assert(EXPECTED_ROOT_MODULE == graphvizText)
+  }
+
+  private fun createProject(name: String): DefaultProject {
+    val project = ProjectBuilder.builder()
+      .withName(name)
+      .withParent(rootProject)
+      .build() as DefaultProject
+
+    project.plugins.apply(JavaLibraryPlugin::class.java)
+    return project
+  }
+}

--- a/plugin/src/test/kotlin/com/jraska/module/graph/assertion/GradleDependencyGraphFactoryTest.kt
+++ b/plugin/src/test/kotlin/com/jraska/module/graph/assertion/GradleDependencyGraphFactoryTest.kt
@@ -9,6 +9,10 @@ import org.junit.Test
 
 class GradleDependencyGraphFactoryTest {
 
+  private val EXPECTED_SINGLE_MODULE = """digraph G {
+":core"
+}"""
+
   private val EXPECTED_GRAPHVIZ = """digraph G {
 ":app" -> ":lib"
 ":app" -> ":feature" [color=red style=bold]
@@ -26,6 +30,7 @@ class GradleDependencyGraphFactoryTest {
 }"""
 
   private lateinit var appProject: DefaultProject
+  private lateinit var coreProject: DefaultProject
   private var rootProject: DefaultProject? = null
 
   @Before
@@ -40,7 +45,7 @@ class GradleDependencyGraphFactoryTest {
     appProject.dependencies.add("implementation", featureProject)
     featureProject.dependencies.add("implementation", libProject)
 
-    val coreProject = createProject("core")
+    coreProject = createProject("core")
     featureProject.dependencies.add("api", coreProject)
     libProject.dependencies.add("implementation", coreProject)
 
@@ -63,6 +68,14 @@ class GradleDependencyGraphFactoryTest {
 
     val graphvizText = GraphvizWriter.toGraphviz(dependencyGraph)
     assert(EXPECTED_TEST_IMPLEMENTATION == graphvizText)
+  }
+
+  @Test
+  fun generatesSingleModuleGraphOnNoDependencyModule() {
+    val dependencyGraph = GradleDependencyGraphFactory.create(coreProject, Api.API_IMPLEMENTATON_CONFIGURATIONS)
+
+    val graphvizText = GraphvizWriter.toGraphviz(dependencyGraph)
+    assert(EXPECTED_SINGLE_MODULE == graphvizText)
   }
 
   private fun createProject(name: String): DefaultProject {

